eDMA loop sg mode and I2S LPUART driver

Raymond0225 opened the pull request:
This patch is addressing these issues:

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/58888
Fixes https://github.com/zephyrproject-rtos/zephyr/issues/70777
Fixes https://github.com/zephyrproject-rtos/zephyr/issues/78291

Tests done:
I2S speed test on RT1170 EVK
uart async api test on RT1170 EVK with bandrate 115200, 1000000 and 2000000bps

Note that:
Please set CONFIG_TEST_USERSPACE=n when you do test on RT1170 EVK or
you will get "write access denied" issue.  It is another issue not
related to LPUART and eDMA drivers.

SynchronicIT commented:
For the LPUART driver, in our real-life application, we tested the proposed solution and noticed an instability. It depends on the TCD queue size. When a certain amount "x" of buffers is consumed and the next buffer should be used, the LPUART or DMA simply stops. 

CONFIG_DMA_TCD_QUEUE_SIZE=2 gave 4 working buffer updates 
CONFIG_DMA_TCD_QUEUE_SIZE=3 gave 5 working buffer updates 
CONFIG_DMA_TCD_QUEUE_SIZE=4 gave 1 working buffer updates (just the initial one..)
CONFIG_DMA_TCD_QUEUE_SIZE=5 gave 3 working buffer updates 
CONFIG_DMA_TCD_QUEUE_SIZE=6 gave 1 working buffer updates (just the initial one..)
CONFIG_DMA_TCD_QUEUE_SIZE=7 gave 2 working buffer updates
...
CONFIG_DMA_TCD_QUEUE_SIZE=16 gave 5 working buffer updates

And although it seems a bit random, I tried each value at least 5 times and every time exact the same amount, although the uart input was random (at 1mbit/s in this case). 

I hope this rings a bell...

matt-wood-ct commented:
Possibly silly question, does a similar issue apply to the lpspi implementation? I've been having a lot of trouble getting that writing properly with dma too. In my case I'm seeing junk data coming out the spi when using it with dma channels configured

hakehuang commented on drivers/serial/uart_mcux_lpuart.c:
```diff
@@ -616,37 +616,22 @@ static int uart_mcux_lpuart_dma_replace_rx_buffer(const struct device *dev)
 	const struct mcux_lpuart_config *config = dev->config;
 	LPUART_Type *lpuart = config->base;
-	struct mcux_lpuart_rx_dma_params *rx_dma_params = &data->async.rx_dma_params;
+	// struct mcux_lpuart_rx_dma_params *rx_dma_params = &data->async.rx_dma_params;
```
why remove this, yo actually change the rx_dma_params

Raymond0225 replied:
I did not change it.  next_rx_buffer is a member of data->async

hakehuang commented on drivers/serial/uart_mcux_lpuart.c:
```diff
@@ -873,6 +865,7 @@ static int mcux_lpuart_rx_buf_rsp(const struct device *dev, uint8_t *buf, size_t
 	assert(data->async.next_rx_buffer_len == 0);
 	data->async.next_rx_buffer = buf;
 	data->async.next_rx_buffer_len = len;
+    uart_mcux_lpuart_dma_replace_rx_buffer(dev);
```
you code indent is wrong should be 8

Raymond0225 replied:
thanks, I will reformat all the related indent and whitespaces.

hakehuang commented on drivers/serial/uart_mcux_lpuart.c:
```diff
@@ -616,37 +616,22 @@ static int uart_mcux_lpuart_dma_replace_rx_buffer(const struct device *dev)
 		dma_reload(config->rx_dma_config.dma_dev, config->rx_dma_config.dma_channel,
-			   LPUART_GetDataRegisterAddress(lpuart), (uint32_t)rx_dma_params->buf,
-			   rx_dma_params->buf_len);
+			   LPUART_GetDataRegisterAddress(lpuart), (uint32_t)data->async.next_rx_buffer,
```
you change the logic: here:
old:
dma_reload next buffer, request buffer update.
now:
request buffer update, so the next is replaced before use, and then you dma_reload the net buffer, which means the original next buffer is not used at all.

Raymond0225 replied:
yes, I reverted part of the code in your last PR. it is because we have to remain buf/buf_len until DMA callback, at the same time, we need submit next_rx_buf to DMA.  The original logic (before your PR  https://github.com/zephyrproject-rtos/zephyr/pull/79325/files) works well. So I use it as before.

DerekSnell commented:
Hi @Raymond0225 ,
I used your eDMA and I2S driver update to test on the FRDM-MCXN947 board.  This is based on the open PR [Add mcxn10 sai module support](https://github.com/zephyrproject-rtos/zephyr/pull/77534).  I had to modify your eDMA driver to support v4, and shared [my branch](https://github.com/nxp-upstream/zephyr/tree/eDMA_loop_SG_N947).  But the `i2s_speed` test passes on FRDM-MCXN947 board in this branch.

Thanks

Raymond0225 commented:
> Hi @Raymond0225 , I used your eDMA and I2S driver update to test on the FRDM-MCXN947 board. This is based on the open PR [Add mcxn10 sai module support](https://github.com/zephyrproject-rtos/zephyr/pull/77534). I had to modify your eDMA driver to support v4, and shared [my branch](https://github.com/nxp-upstream/zephyr/tree/eDMA_loop_SG_N947). But the `i2s_speed` test passes on FRDM-MCXN947 board in this branch.
> 
> Thanks

thanks Derek.  I wrote the code for DMA V4 based on an old version. I will update the code based on your modifications.

Raymond0225 commented:
> For the LPUART driver, in our real-life application, we tested the proposed solution and noticed an instability. It depends on the TCD queue size. When a certain amount "x" of buffers is consumed and the next buffer should be used, the LPUART or DMA simply stops.
> 
> CONFIG_DMA_TCD_QUEUE_SIZE=2 gave 4 working buffer updates CONFIG_DMA_TCD_QUEUE_SIZE=3 gave 5 working buffer updates CONFIG_DMA_TCD_QUEUE_SIZE=4 gave 1 working buffer updates (just the initial one..) CONFIG_DMA_TCD_QUEUE_SIZE=5 gave 3 working buffer updates CONFIG_DMA_TCD_QUEUE_SIZE=6 gave 1 working buffer updates (just the initial one..) CONFIG_DMA_TCD_QUEUE_SIZE=7 gave 2 working buffer updates ... CONFIG_DMA_TCD_QUEUE_SIZE=16 gave 5 working buffer updates
> 
> And although it seems a bit random, I tried each value at least 5 times and every time exact the same amount, although the uart input was random (at 1mbit/s in this case).
> 
> CONFIG_DMA_TCD_QUEUE_SIZE=1 actually gave the most stable result, and it looks like your initializing more buffers than needed at: zephyr/drivers/dma/dma_mcux_edma.c:391 (the max queue size instead of the requested block_count...
> 
> I hope this rings a bell...

Thanks for your feedback. I don't understand what you said. I do use all TCDs and linked them together. Can you elaborate how to reproduce the "stops" issue you mentioned here?

dleach02 commented:
@Raymond0225 please clean up the file changes to remove the spaces at beginning of lines. The Zephyr coding standard is to use tabs.

Raymond0225 commented:
> @Raymond0225 please clean up the file changes to remove the spaces at beginning of lines. The Zephyr coding standard is to use tabs.

thanks. I am cleaning up the code, will update the PR later.

SynchronicIT commented:
> Thanks for your feedback. I don't understand what you said. I do use all TCDs and linked them together. Can you elaborate how to reproduce the "stops" issue you mentioned here?

We are building a uart to usb bridge, similar to the usb example. As we need to switch both the DTR and baud rate in real-time, we're using the next-gen usb driver (and have made some fixes for it to make it more robust). 
To get the lowest latency, we are coupling the usb uart and physical uart in the irq and async callback functions. 
On other chipset's we used solely the fifo / irq method, but due to the small fifo on the RT1060, we need to use the async/dma. 

Concerning the Rx and TCDs: 
I'm feeding the physical uart now with a data set every 1[s] of 256 bytes at 1mbit and 2 mbit and provide the async driver with buffers of 2KB. If I use anything else than a single TCD by setting it in the config (which reflects the .block_count == 1) I do get several requests for new buffers and buffer release events. But after the mentioned number the requests and releases stop, but any irq from the DMA stops. The uart irq keeps on running (giving overflow errors of course). It reproduced very well for the numbers given, which suggests there is a good relationship. 
I don't have a simplified example to test it with, which makes it a bit difficult to hand it over. 

Concerning  the Tx side: 
Having the Tx side from a uart_fifo_irq (of the usb) directly towards the physical uart using the async. (thus not using a separate thread to handle the copy / forwarding with the goal to minimize the latency) The Tx Side of the async seems to be giving inconsistent data output, as with too much data, the uart output becomes 10 bits '0' ( a break instead of start - 8 bytes - stop). I've added atomic bits and I have a "uart_tx(..)" in the TX_READY event handling. I'm not sure if that is even allowed. 

Concerning  switching the baudrate: 
When switching the baud rate, the lpuart driver completely reinitializes the driver. In case the DMA is still active, re-enabling the rx dma (uart_rx(...)) fails as the DMA is still active. I added code to forcefully stop the DMA before calling the configure and restart the uart_rx dma. I see two problems. The baud-rate doesn't seem to switch properly (as the data is partially received (although the baudrate is a factor 4 lower), and the buffers are not flushed ( dma seems to continue where it was previously). 


Perhaps we could take this discussion offline / beyond this github issue / pull-request.

mmahadevan108 commented on drivers/dma/dma_mcux_edma.c:
Please squash this commit with the earlier commit that changes this file.

Raymond0225 replied:
thanks for your feedback. I am not familiar with git commands, I will try to it.

Raymond0225 commented:
> > Thanks for your feedback. I don't understand what you said. I do use all TCDs and linked them together. Can you elaborate how to reproduce the "stops" issue you mentioned here?
> 
> We are building a uart to usb bridge, similar to the usb example. As we need to switch both the DTR and baud rate in real-time, we're using the next-gen usb driver (and have made some fixes for it to make it more robust). To get the lowest latency, we are coupling the usb uart and physical uart in the irq and async callback functions. On other chipset's we used solely the fifo / irq method, but due to the small fifo on the RT1060, we need to use the async/dma.
> 
> Concerning the Rx and TCDs: I'm feeding the physical uart now with a data set every 1[s] of 256 bytes at 1mbit and 2 mbit and provide the async driver with buffers of 2KB. If I use anything else than a single TCD by setting it in the config (which reflects the .block_count == 1) I do get several requests for new buffers and buffer release events. But after the mentioned number the requests and releases stop, but any irq from the DMA stops. The uart irq keeps on running (giving overflow errors of course). It reproduced very well for the numbers given, which suggests there is a good relationship. I don't have a simplified example to test it with, which makes it a bit difficult to hand it over.
> 
> Concerning the Tx side: Having the Tx side from a uart_fifo_irq (of the usb) directly towards the physical uart using the async. (thus not using a separate thread to handle the copy / forwarding with the goal to minimize the latency) The Tx Side of the async seems to be giving inconsistent data output, as with too much data, the uart output becomes 10 bits '0' ( a break instead of start - 8 bytes - stop). I've added atomic bits and I have a "uart_tx(..)" in the TX_READY event handling. I'm not sure if that is even allowed.
> 
> Concerning switching the baudrate: When switching the baud rate, the lpuart driver completely reinitializes the driver. In case the DMA is still active, re-enabling the rx dma (uart_rx(...)) fails as the DMA is still active. I added code to forcefully stop the DMA before calling the configure and restart the uart_rx dma. I see two problems. The baud-rate doesn't seem to switch properly (as the data is partially received (although the baudrate is a factor 4 lower), and the buffers are not flushed ( dma seems to continue where it was previously).
> 
> Perhaps we could take this discussion offline / beyond this github issue / pull-request.

thanks for your feedback. I suggest you have a look at the the test code of uart_async_api test case, there is a function "test_double_buffer" which may be close to your use case.  block_count is only valid during DMA configuration. I suggest you set it as 1. after that, each time when driver request user rx buffer, callback function will be triggered. Driver would not ask for new buffer until current rx buffer is filled fully. There may be multi rx ready event between 2 buf requests.

SynchronicIT commented:
> thanks for your feedback. I suggest you have a look at the the test code of uart_async_api test case, there is a function "test_double_buffer" which may be close to your use case. block_count is only valid during DMA configuration. I suggest you set it as 1. after that, each time when driver request user rx buffer, callback function will be triggered. Driver would not ask for new buffer until current rx buffer is filled fully. There may be multi rx ready event between 2 buf requests. 2 TCDs usually is enough. It is not related to block_count. I suggest set block_count to 1 whatever how many TCD you set.


I've ran the proposed tests. The test succeeds  only when using the TCM regions for the buffers. Is this mandatory? 
If the baudrate is higher than 115200 the first tests fail. Also the tests only simulate ideal situation. (Only transmit when the Rx is handled completely). . During the tests i do see the error "No buffers to release from RX DMA!"
I'll update the buffers to reside in the TCM to see if that solves my challenges. 

Tx seems to be working now, by moving "uart_tx(..)" to a work task instead of directly in the callback. This is aparently not allowed. (perhaps something to add to the documentation)

Switching baudrate at runtime while DMA is active is still a challenge. Not sure what is actually failing (the dma buffers or the uart settings). Using the fifo api switching baudrate works well.

hakehuang commented:
@Raymond0225 , I rebase you pr to the latest zephyr code. and test on mimxrt1170_evk_cm7 it fails as below log

scripts/twister --device-testing --device-serial /dev/ttyACM0 -T tests/drivers/uart/uart_async_api/

```
uart:~$ *** Booting Zephyr OS build (tainted) v4.0.0-rc1-15-g359a81e7db80 ***
Running TESTSUITE uart_async_chain_read
===================================================================
E: No buffers to release from RX DMA!
E: No buffers to release from RX DMA!
START - test_chained_read

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/uart/uart_async_api/src/test_uart_async.c:355: test_chained_read_callback: rx_data_idx + evt->data.rx.len <= sizeof(chained_cpy_buf) is false

ASSERTION FAIL [0] @ WEST_TOPDIR/zephyr/kernel/sched.c:1393
        aborting essential thread 0x80002580
E: r0/a1:  0x00000004  r1/a2:  0x00000571  r2/a3:  0x4007c000
E: r3/a4:  0x00000004 r12/ip:  0x00000000 r14/lr:  0x3000a55b
E:  xpsr:  0x21000000
E: Faulting instruction address (r15/pc): 0x3000b858
E: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
E: Current thread: 0x80002580 (sysworkq)
E: Halting system

```

DerekSnell commented:
Hi @SynchronicIT ,
> The test succeeds only when using the TCM regions for the buffers. Is this mandatory? 

The buffers used by the DMA need to be in non-cached memory, otherwise there may be a cache coherency issue.  The TCMs are not cached, which is one option for the buffers.  Another option is to use slower memories like external SDRAM or internal OCRAM, but the buffers need to be placed in a `nocache` region.  On the RT1xxx platform, the cache is only used by the CPU.  Other masters in the SOC like the DMA do not use the cache.  This can lead to a cache coherency issue where the DMA is reading/writing to the physical RAM, but the CPU is accessing the cache.

> Switching baudrate at runtime while DMA is active is still a challenge. 

Your use-case to dynamically change the baud rate is beyond the scope of this PR.  You can create a [GitHub Discussion](https://github.com/zephyrproject-rtos/zephyr/discussions) for that separate topic.

Thanks for your feedback with this PR.

DerekSnell commented on drivers/i2s/i2s_mcux_sai.c:
Hi @Raymond0225 ,
@hakehuang raised a great point offline.  The changes here to i2s_mcux_sai.c do not set the `cyclic` bit.  It appears this driver is not using the new loop SG feature in the eDMA driver.  Is that intended?

Raymond0225 replied:
No, I forgot to enable it. will fix it.

Raymond0225 commented:
> @Raymond0225 , I rebase you pr to the latest zephyr code. and test on mimxrt1170_evk_cm7 it fails as below log
> 
> scripts/twister --device-testing --device-serial /dev/ttyACM0 -T tests/drivers/uart/uart_async_api/
> 
> ```
> uart:~$ *** Booting Zephyr OS build (tainted) v4.0.0-rc1-15-g359a81e7db80 ***
> Running TESTSUITE uart_async_chain_read
> ===================================================================
> E: No buffers to release from RX DMA!
> E: No buffers to release from RX DMA!
> START - test_chained_read
> 
>     Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/uart/uart_async_api/src/test_uart_async.c:355: test_chained_read_callback: rx_data_idx + evt->data.rx.len <= sizeof(chained_cpy_buf) is false
> 
> ASSERTION FAIL [0] @ WEST_TOPDIR/zephyr/kernel/sched.c:1393
>         aborting essential thread 0x80002580
> E: r0/a1:  0x00000004  r1/a2:  0x00000571  r2/a3:  0x4007c000
> E: r3/a4:  0x00000004 r12/ip:  0x00000000 r14/lr:  0x3000a55b
> E:  xpsr:  0x21000000
> E: Faulting instruction address (r15/pc): 0x3000b858
> E: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
> E: Current thread: 0x80002580 (sysworkq)
> E: Halting system
> ```
> 
> last commit after rebase ''' commit 359a81e7db8074487a1d85bf059848a6e92bfac4 (HEAD -> pr_80584_edma_async) Author: Raymond [raymond.lei@nxp.com](mailto:raymond.lei@nxp.com) Date: Wed Oct 30 15:31:19 2024 -0500
> 
> ```
> drivers: dma: edma: Fix compile issue for eDMA V4
> 
> Some data structures are different between eDMA and eDMA V4, use
> different macros defined in SDK driver for each of them.
> Make indent/leading space modifications.
> 
> Signed-off-by: Raymond <raymond.lei@nxp.com>
> ```
> 
> '''

Thanks Hake. I got same error when I try to put TCD pool to DTCM. it is because some fields inside edma_tcd_t is not initialized.
I will fix it in the following commit.  I add a line "memset(&DEV_CFG(dev)->tcdpool[channel][i],0,sizeof(DEV_CFG(dev)->tcdpool[channel][i]));" during TCD pool initialization.

danieldegrasse commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -358,26 +374,93 @@ static int dma_mcux_edma_configure(const struct device *dev, uint32_t channel,
+		data->transfer_settings.write_idx = 0;
+		data->transfer_settings.empty_tcds = CONFIG_DMA_TCD_QUEUE_SIZE;
+
+		EDMA_PrepareTransfer(
```
I'm unclear why we prepare a transfer here, and then configure each TCD using it. It seems like we will overwrite the `SADDR`, `DADDR`, `CITER`, and `BITER` fields later on in the loop below. Why not just initialize the TCDs directly in that loop? We could set `DLAST_SGA` directly as well in the loop below, rather than calling `EDMA_TcdSetTransferConfig`

Raymond0225 replied:
Just want to use this API in SDK driver instead of writing a new function.
yes, we can write a new function to do it as well.

danieldegrasse commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -507,29 +614,119 @@ static int dma_mcux_edma_reload(const struct device *dev, uint32_t channel,
+#ifdef CONFIG_DMA_MCUX_EDMA
+		while(DEV_BASE(dev)->TCD[channel].CSR & DMA_CSR_ACTIVE_MASK)
+#elif (CONFIG_DMA_MCUX_EDMA_V3 || CONFIG_DMA_MCUX_EDMA_V4)
+		while(DEV_BASE(dev)->CH[channel].TCD_CSR & DMA_CH_CSR_ACTIVE_MASK)
```
I'm not sure I'm following the logic here- why do we wait until the EDMA is not active here? The `dma_reload` function should simply queue a new buffer into the EDMA (which occurs above) then exit, right? There is no requirement to wait for the DMA to complete execution here.

It seems like at this point we should:
- clear the DREQ bit on the prior TCD, so the next TCD can be loaded
- check the next TCD address, and restart the DMA if `hw_id == sw_id` (as this indicates the TCD we just configured was loaded already, right?)
- otherwise, set the ESG bit to resolve the HW bug, and make sure the DREQ bit is set on the TCD we are queueing

I'm not clear why we need to disable channel requests or wait for the EDMA to be inactive. Shouldn't it be sufficient to configure the TCD, enable the EDMA to load the TCD, then perform the safety check to make sure the EDMA did not load an invalid TCD?

Raymond0225 replied:
If we do not stop DMA during this "appending procedure", HW ID may change at any time, that is to say, we don't know exactly which TCD DMA is working on.  DMA may have finished all transfers, may be transferring the last one, may be transferring the others, all handling of these conditions are different.
By disable the channel request, we cut off the signal from the peripheral, no more transfer request. DMA stops after current burst finished. Why we wait for DMA to be inactive is because even we cut off the transfer request, there may be a burst transfer is on-going, DMA has its own internal buffer and logics. DMA may modify the same HW reg while CPU is trying to make a modification.

danieldegrasse replied:
Ok, that makes sense. Isn't there a risk here though of "drops" in transfers? For example with audio, wouldn't the DMA being inactive here manifest as a momentary pause in the I2S audio stream and therefore a pause in the audio from the codec?

Raymond0225 replied:
Risk of "drops":
that's why we need make this part of code as short as possible. The longest path of code between EDMA_DisableChannelRequest and EDMA_EnableChannelRequest is in case DMA has finished all transfers and we need rewrite HW regs. Even in this case, all codes to execute are about 10 simple instructions(set value, comparing).  It should be very quick.  On RT1170, SAI has 16 bytes FIFO, given the sample rate: 44K, 2 bytes word, 2 channels, it takes about 90us to full the FIFO.  I think it is much far longer than the execution time of the critical code.

last TCD transfering:
In this case, hw_id should be current writing TCD address, we have code to handle this case:
				/* DMA is running on last transfer. HW has loaded the last one,
				 * we need ensure it's DREQ is cleared.
				 */
				EDMA_EnableAutoStopRequest(DEV_BASE(dev), channel, false);
				LOG_DBG("Last transfer.");_

Raymond0225 commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -254,6 +266,9 @@ static int dma_mcux_edma_configure(const struct device *dev, uint32_t channel,
 	int ret = 0;
+	edma_tcd_t *tcd = NULL;
+
+	tcd = &(DEV_CFG(dev)->tcdpool[channel][data->transfer_settings.write_idx]);
```
no, we don't need initialize tcd here, this line is a typo, I will correct it. thanks

Raymond0225 commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -358,26 +374,93 @@ static int dma_mcux_edma_configure(const struct device *dev, uint32_t channel,
+		while (block_config != NULL && data->transfer_settings.write_idx < CONFIG_DMA_TCD_QUEUE_SIZE) {
+
+			tcd = &(DEV_CFG(dev)->tcdpool[channel][data->transfer_settings.write_idx]);
+#if (CONFIG_DMA_MCUX_EDMA || CONFIG_DMA_MCUX_EDMA_V3)
```
good idea, let me try to improve this part of code.

Raymond0225 commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -358,26 +374,93 @@ static int dma_mcux_edma_configure(const struct device *dev, uint32_t channel,
+				tcd->CSR &= ~DMA_CSR_DREQ(1U);
+			}
+#elif CONFIG_DMA_MCUX_EDMA_V4
+			EDMA_TCD_SADDR(tcd,kEDMA_EDMA4Flag) = block_config->source_address;
```
Good idea

Raymond0225 commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -358,26 +374,93 @@ static int dma_mcux_edma_configure(const struct device *dev, uint32_t channel,
+			block_config = block_config->next_block;
+
+
+		}
```
good idea

Raymond0225 commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -492,10 +575,34 @@ static int dma_mcux_edma_resume(const struct device *dev, uint32_t channel)
 }
 
 
+static void dma_mcux_edma_update_hw_tcd(const struct device *dev, uint32_t channel,
```
have done on new commit ,thanks for your suggestions.

Raymond0225 commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -507,29 +614,119 @@ static int dma_mcux_edma_reload(const struct device *dev, uint32_t channel,
+		tcd = &(DEV_CFG(dev)->tcdpool[channel][data->transfer_settings.write_idx]);
+		pre_tcd = &(DEV_CFG(dev)->tcdpool[channel][(data->transfer_settings.write_idx - 1)%CONFIG_DMA_TCD_QUEUE_SIZE]);
+#if (CONFIG_DMA_MCUX_EDMA || CONFIG_DMA_MCUX_EDMA_V3)
+		tcd->SADDR = src;
```
done

Raymond0225 commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -507,29 +614,119 @@ static int dma_mcux_edma_reload(const struct device *dev, uint32_t channel,
+
+		/* Use DLAST_SGA as the HW ID */
+		hw_id = EDMA_GetNextTCDAddress(DEV_EDMA_HANDLE(dev,channel));
+		if (data->transfer_settings.empty_tcds >= CONFIG_DMA_TCD_QUEUE_SIZE || hw_id == sw_id) {
```
yes, I think >=  is better than == :)

DerekSnell reviewed (changes requested):
Hi @Raymond0225 ,
Thank you for all your hard work with these fixes.  

I am only blocking this to see your updates to i2s_mcux_sai.c with the loop SG support.

Otherwise, I see no issues with your updates.  I do think we can improve with comments to help understand how your eDMA driver works, and provided some comments for you to review.

Also, regarding this note in your commit message:
> please set CONFIG_TEST_USERSPACE=n when you do test on RT1170 EVK or
you will get "write access denied" issue.  It is another issue not related to LPUART and eDMA drivers.

It's good to include that detail in the PR description (and I see it is already there), so others are aware how to replicate your test.  But I would not include in the commit message.  Over time, that `CONFIG_TEST_USERSPACE=n` may no longer apply.

And for the PR description, I would point out the `CONFIG_TEST_USERSPACE=n` requirement is specifically for the uart_async_api test.  The drivers in this PR are not impacted by that Kconfig, but the test app for the LPUART driver will fail today with userspace enabled.

Thanks

Raymond0225 committed:
drivers: dma: edma: Put the TCD pool in DTCM by default

To reduce the latency of CPU accessing/modifying SW TCD content, it
better to put TCD pool to DTCM by default.

Signed-off-by: Raymond Lei <raymond.lei@nxp.com>

Raymond0225 commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -358,26 +374,93 @@ static int dma_mcux_edma_configure(const struct device *dev, uint32_t channel,
-				LOG_ERR("Error submitting EDMA Transfer: 0x%x", submit_status);
-				ret = -EFAULT;
+			LOG_ERR("Error submitting EDMA Transfer: 0x%x", submit_status);
+			ret = -EFAULT;
```
These are modifications suggested by the check_compliance script, if i do not make the changes, it will fail the check.

Raymond0225 commented on drivers/i2s/i2s_mcux_sai.c:
ok

Raymond0225 commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -181,7 +188,12 @@ static void nxp_edma_callback(edma_handle_t *handle, void *param, bool transferD
+		data->transfer_settings.empty_tcds++;
+		/*In loop mode, DMA is always busy*/
+		data->busy = 1;
+		ret = DMA_STATUS_COMPLETE;
```
I think we should return "complete" instead of "block". Each time when callback is triggered , it indicate current transfer is completed.

Raymond0225 commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -507,29 +614,119 @@ static int dma_mcux_edma_reload(const struct device *dev, uint32_t channel,
+#endif
+		{
+			;
+		}
```
I have submitted new implementation. It passed the compliance check. "while(xxx);" is not recommended by the compliance script.

Raymond0225 commented:
> Hi @Raymond0225 , Thank you for all your hard work with these fixes.
> 
> I am only blocking this to see your updates to i2s_mcux_sai.c with the loop SG support.
> 
> Otherwise, I see no issues with your updates. I do think we can improve with comments to help understand how your eDMA driver works, and provided some comments for you to review.
> 
> Also, regarding this note in your commit message:
> 
> > please set CONFIG_TEST_USERSPACE=n when you do test on RT1170 EVK or
> > you will get "write access denied" issue.  It is another issue not related to LPUART and eDMA drivers.
> 
> It's good to include that detail in the PR description (and I see it is already there), so others are aware how to replicate your test. But I would not include in the commit message. Over time, that `CONFIG_TEST_USERSPACE=n` may no longer apply.
> 
> And for the PR description, I would point out the `CONFIG_TEST_USERSPACE=n` requirement is specifically for the uart_async_api test. The drivers in this PR are not impacted by that Kconfig, but the test app for the LPUART driver will fail today with userspace enabled.
> 
> Thanks

@DerekSnell  Thanks for your comments. I submitted a commit to include all these modifications(only comments and copyright, no logic/function modifications).

DerekSnell commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -357,26 +399,93 @@ static int dma_mcux_edma_configure(const struct device *dev, uint32_t channel,
+					&DEV_CFG(dev)->tcdpool[channel][(i + 1) %
+									CONFIG_DMA_TCD_QUEUE_SIZE]);
+
+				/* Enable Major interrupt.*/
```
```suggestion
				/* Enable Major loop interrupt.*/
```

DerekSnell commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -357,26 +399,93 @@ static int dma_mcux_edma_configure(const struct device *dev, uint32_t channel,
+				block_config = block_config->next_block;
+			}
+
+			if (block_config && data->transfer_settings.empty_tcds == 0) {
```
Should this be:
```suggestion
			if ((block_config != NULL) &&
				(data->transfer_settings.empty_tcds == 0)) {
```

DerekSnell reviewed (changes requested):
Much improved, @Raymond0225.  And easier to review now, thank you.

But it still needs a little work.

Raymond0225 commented on drivers/dma/Kconfig.mcux_edma:
```diff
@@ -44,6 +44,7 @@ config DMA_MCUX_TEST_SLOT_START
 
 config DMA_MCUX_USE_DTCM_FOR_DMA_DESCRIPTORS
 	bool "Use DTCM for DMA descriptors"
+	default y
```
thanks, I am looking into this failure.

hakehuang commented:
now this PR works on my boards with uart_async_api and edma tests. @Raymond0225  can you enable the i2s on mimxrt1170_evk board in this PR, so that I can run a test with this cyclic dma with i2s.

Raymond0225 commented:
> now this PR works on my boards with uart_async_api and edma tests. @Raymond0225 can you enable the i2s on mimxrt1170_evk board in this PR, so that I can run a test with this cyclic dma with i2s.

To run i2s_speed test on RT1170 EVK, you need HW reworks:

Board-specific details:

MIMXRT1170_EVK:
This board uses CONFIG_I2S_TEST_SEPARATE_DEVICES=y and connects two SAI peripherals by shorting
signals externally on the EVK.  These are the HW changes required to run this test:
	- Remove jumper J8 and resistor R78
	- Short BCLK J9-pin1  (SAI1_RX_BCLK)  to J66-pin1  (SAI4_TX_BCLK)
	- Short SYNC J9-pin5  (SAI1_RX_SYNC)  to J64-pin1  (SAI4_TX_SYNC)
	- Short Data J61-pin1 (SAI1_RX_DATA)  to J63-pin1  (SAI4_TX_DATA)

Raymond0225 committed:
drivers: dma: edma: Pending length calculation incorrect

Pending length should be mainloop count multiply NBYTES of minor loops.

Signed-off-by: Raymond Lei <raymond.lei@nxp.com>

Raymond0225 committed:
drivers: dma: edma: eDMA loop SG mode support

Existing dynamic SG mode have some issues which cause uart sync api and
I2S speed test failed.

These issues are: a wrong "Done" bit issue found on UART async api
test. An invalid destination address issue found on I2S speed
test. By introducing loop SG mode, these issues are all fixed.

Some data structures are different between eDMA and eDMA V4, use
different macros defined in SDK driver for each of them.

Signed-off-by: Raymond Lei <raymond.lei@nxp.com>

Raymond0225 committed:
Drivers: driver: i2s: Use eDMA loop SG mode

I2S speed test failed because of "Invalid destination address" issue By use
eDMA loop SG mode, this issue is fixed.

Fixes: #70777

Signed-off-by: Raymond Lei <raymond.lei@nxp.com>

Raymond0225 committed:
drivers: serial: lpuart: Enable SG mode of LPUART

New implemented eDMA loop SG mode can be used to lpuart driver to fix
the conflict of "Done" bit issue which cause uart_async_api test failed
when SG mode is enabled.

Also, this loop SG mode works well even with a high bandrate(test done
on 2000000bps).
Fixes: #78291

Signed-off-by: Raymond Lei <raymond.lei@nxp.com>

DerekSnell reviewed (approved):
Great work, @Raymond0225 .  Thanks for all your hard work with this.

danieldegrasse commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -507,29 +625,115 @@ static int dma_mcux_edma_reload(const struct device *dev, uint32_t channel,
+		EDMA_ClearChannelStatusFlags(DEV_BASE(dev), channel, kEDMA_DoneFlag);
+		EDMA_HW_TCD_CSR(dev, channel) |= DMA_CSR_ESG_MASK;
+#elif (CONFIG_DMA_MCUX_EDMA_V3 || CONFIG_DMA_MCUX_EDMA_V4)
+		/*We have not verified if this issue exist on V3/V4 HW, jut place a holder here. */
```
```suggestion
		/*We have not verified if this issue exist on V3/V4 HW, just place a holder here. */
```

danieldegrasse reviewed (approved):
A few minor (nonblocking) nits, overall change looks good to me though. Thanks for your work here!

RadekPolyend commented:
This PR breaks SAI driver

DerekSnell commented:
Hi @RadekPolyend ,
This PR was tested with the [i2s_speed test](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/i2s/i2s_speed) on the MIMXRT1170-EVK board.  

Can you share more details how to replicate the issue you are seeing?  I recommend creating a new [Issue](https://github.com/zephyrproject-rtos/zephyr/issues/new/choose), and adding details there to replicate and understand the root cause.  You can link that issue to this PR.

Thanks for raising this.  Best regards

RadekPolyend commented:
Ok but the test doesn't seem to assume the reload of the tx buffers after the start trigger, it just drains the buffers which were loaded before the trigger

RadekPolyend commented:
https://github.com/zephyrproject-rtos/zephyr/issues/82470
